### PR TITLE
AST: add `Format` interface on `ExprNode` for convert AST back to string

### PR DESCRIPTION
**Pull request by hicqu**
In many cases we needs to convert AST back to string. Use the raw text maybe is a bad idea because that's not pretty when print. So we add this interface.

### CODE REVIEW

**Comment by hicqu** [comment]
@coocood @XuHuaiyu PTAL thanks.

**Comment by shenli** [comment]
@hicqu Is there any example that the raw text could not satisfy the need?

File: ast/expressions.go [hunk 1]
```diff
@@ -543,6 +640,16 @@ type PatternLikeExpr struct {
 	PatTypes []byte
 }
 
+func (n *PatternLikeExpr) Format(w io.Writer) {
+	n.Expr.Format(w)
+	w.Write(hack.Slice(" LIKE "))
+	n.Pattern.Format(w)
+	if n.Escape != '\\' {
+		w.Write(hack.Slice(" ESCAPE "))
+		w.Write(hack.Slice(fmt.Sprintf("'%c'", n.Escape)))
```
Comment (coocood) [inline ast/expressions.go]: `fmt.Fprintf` will do.

**Comment by hicqu** [comment]
@shenli For example, if users write SQL like this:
```sql
CREATE TABLE x (
    a int primary key,
    b int as (  a +          8       ) -- bad format in generation expression.
);
```
If we use the raw text directly, we can't print it pretty in `show create table x`. So we need get the string from AST to remove all redundant blanks.

File: ast/expressions.go [hunk 2]
```diff
@@ -373,6 +426,11 @@ type ColumnNameExpr struct {
 	Refer *ResultField
 }
 
+// Format the ExprNode into a Writer.
+func (n *ColumnNameExpr) Format(w io.Writer) {
+	fmt.Fprintf(w, n.Name.String())
```
Comment (shenli) [inline ast/expressions.go]: Seems MySQL will add single-quote around the column name:

> mysql> CREATE TABLE x (
>     ->     a int primary key,
>     ->     b int as (  a +          8       ) -- bad format in generation expression.
>     -> );
> Query OK, 0 rows affected (0.09 sec)
> 
> mysql> show create table x;
> +-------+------------------------------------------------------------------------------------------------------------------------------------------------------------+
> | Table | Create Table                                                                                                                                               |
> +-------+------------------------------------------------------------------------------------------------------------------------------------------------------------+
> | x     | CREATE TABLE `x` (
>   `a` int(11) NOT NULL,
>   `b` int(11) GENERATED ALWAYS AS ((`a` + 8)) VIRTUAL,
>   PRIMARY KEY (`a`)
> ) ENGINE=InnoDB DEFAULT CHARSET=utf8 |
> +-------+------------------------------------------------------------------------------------------------------------------------------------------------------------+
> 1 row in set (0.00 sec)
> 
> mysql>

**Comment by shenli** [comment]
@hicqu CI failed.

File: ast/expressions.go [hunk 3]
```diff
@@ -448,6 +516,19 @@ type PatternInExpr struct {
...
 }
 
+// Format the ExprNode into a Writer.
+func (n *PatternInExpr) Format(w io.Writer) {
+	n.Expr.Format(w)
+	fmt.Fprintf(w, " IN (")
+	for i, expr := range n.List {
+		expr.Format(w)
+		if i != len(n.List)-1 {
+			fmt.Fprintf(w, ", ")
```
Comment (XuHuaiyu) [inline ast/expressions.go]: seems there is no blank in MySQL when formatting InExpr.

File: ast/expressions.go [hunk 4]
```diff
@@ -700,6 +850,12 @@ type UnaryOperationExpr struct {
 	V ExprNode
 }
 
+// Format the ExprNode into a Writer.
+func (n *UnaryOperationExpr) Format(w io.Writer) {
+	n.Op.Format(w)
```
Comment (XuHuaiyu) [inline ast/expressions.go]: for `create table t(a int as (+1))`,
MySQL:
```
CREATE TABLE `t` (
  `a` int(11) GENERATED ALWAYS AS (1) VIRTUAL
) ENGINE=InnoDB DEFAULT CHARSET=latin1
```

for `create table t(a int as (-1))`
MySQL:
``` sql
CREATE TABLE `t` (
  `a` int(11) GENERATED ALWAYS AS (-(1)) VIRTUAL
) ENGINE=InnoDB DEFAULT CHARSET=latin1
```

File: ast/format_test.go [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+func getDefaultCharsetAndCollate() (string, string) {
+	return "utf8", "utf8_bin"
+}
+
+func (ts *testAstFormatSuite) TestAstFormat(c *C) {
+	var testcases = []struct {
+		input  string
+		output string
+	}{
+		// Literals.
```
Comment (XuHuaiyu) [inline ast/format_test.go]: add test for UnaryOperation

File: ast/format_test.go [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+		{`350`, `350`},
+		{`345.678`, `345.678`},
```
Comment (XuHuaiyu) [inline ast/format_test.go]: add test for `1e-12`, `1E-12`

File: ast/format_test.go [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+		{`3 is false`, `3 IS FALSE`},
+		{`  ( x is false  )`, `(x IS FALSE)`},
+		{`3 in ( a,b,"h",6 )`, `3 IN (a, b, "h", 6)`},
+		{`"abc" like '%b%'`, `"abc" LIKE "%b%"`},
+		{`"abc" like '%b%' escape '_'`, `"abc" LIKE "%b%" ESCAPE '_'`},
+		{`"abc" regexp '.*bc?'`, `"abc" REGEXP ".*bc?"`},
+		{`"abc" not regexp '.*bc?'`, `"abc" NOT REGEXP ".*bc?"`},
+		{`-  4`, `-4`},
+		// Functions.
+		{` json_extract ( a,'$.b',"$.\"c d\"" ) `, `json_extract(a, "$.b", "$.\"c d\"")`},
```
Comment (XuHuaiyu) [inline ast/format_test.go]: add more test for functions,
like `in`

File: ast/expressions.go (hunk 4 above)
Comment (hicqu) [reply ast/expressions.go]: I don't think here we need to be 100% with MySQL. we don't need to add extra `(` and `)` for neg operator.

**Comment by hicqu** [comment]
@XuHuaiyu PTAL thanks.

**Comment by XuHuaiyu** [comment]
LGTM

**Comment by shenli** [comment]
LGTM

**Pull request merged**